elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,260 @@
+        Simulator fit with the observations of the true/underlying model.
+    fn_distance : str or function
+        Function for calculating distance (can be a string if the function is implemented in ELFI).
+    seed : int, optional
+
+    References
+    ----------
+    [1] Nunes, M. A., & Balding, D. J. (2010).
+    On optimal selection of summary statistics for approximate Bayesian computation.
+    Statistical applications in genetics and molecular biology, 9(1).
+    [2] Blum, M. G., Nunes, M. A., Prangle, D., & Sisson, S. A. (2013).
+    A comparative review of dimension reduction methods in approximate Bayesian computation.
+    Statistical Science, 28(2), 189-208.
+
+    """
+
+    def __init__(self, list_ss, simulator, fn_distance, seed=0):
+        """Initialise the summary-statistics selection based on the Two-Stage Procedure.
+
+        Parameters
+        ----------
+        list_ss : array_like
+            List of the to-be-assessed summary statistics.
+        simulator : elfi.Simulator
+            Simulator fit with the observations of the true/underlying model.
+        fn_distance : str or function
+            Function for calculating distance
+            (can be a string if the function is implemented in ELFI).
+        seed : int, optional
+
+        """
+        self.list_ss = list_ss
+        self.simulator = simulator
+        self.fn_distance = fn_distance
+        self.seed = seed
+
+    def run(self, k, n_sim, n_acc, n_closest):
+        """Run the Two-Stage Procedure.
+
+        Parameters
+        ----------
+        k : int
+            Parameter for the kth-nearest-neighbour search.
+        n_sim : int
+            Number of the total ABC-rejection simulations.
+        n_acc : int
+            Number of the accepted ABC-rejection simulations.
+        n_closest : int
+            Number of the `closest' datasets.
+
+        Returns
+        -------
+        array_like
+            Set of the summary statistics showing the optimal performance.
+
+        """
+        # Obtain the combinations of the summary statistics.
+        sets_candidate_ss = self._create_ss_combinations(max_cardinality=k)
+
+        # Find the summary statistics combination with the minimum entropy,
+        # preserve the parameters (thetas) corresponding to the `closest' datasets.
+        thetas = {}
+        E_me = np.inf
+        for set_ss in sets_candidate_ss:
+            names_ss = [ss.__name__ for ss in set_ss]
+            thetas_ss = self._obtain_accepted_thetas(set_ss, n_sim, n_acc)
+            thetas[set_ss] = thetas_ss
+            E_ss = self._calc_entropy(thetas_ss, n_acc, k)
+            if E_ss < E_me:
+                E_me = E_ss
+                names_ss_me = names_ss
+                thetas_closest = thetas_ss[:n_closest]
+            logger.info('Combination %s shows the entropy of %f' % (names_ss, E_ss))
+        logger.info('\nThe minimum entropy of %f was found in %s.\n' % (E_me, names_ss_me))
+
```

Comment:
Do we want to generalise summary-statistics selection? That is, provide a parent `SummaryStatisticsSelection` interface which summary-statistics selection techniques (such as minimum entropy and Two-Stage Procedure) would follow? In this way, we could easily call say MinimumEntropySelection and TwoStageSelection separately (where TwoStageSelection would have an implicit call to the minimum entropy's algorithm). All this would need  time to think over the design of the diagnostics. However, would it not be more time-efficient to generalise the summary-statistics diagnostics once there is an alternative selection technique set to be implemented in ELFI? As suggested by the review on summary-selection techniques by Blum et al. (2013) there is no advantage in running the minimum-entropy selection alone; on contrary, there are some suggestions to have alternatives for the minimum-entropy selection used in Two-Stage Procedure's Stage 1 (which was proposed in 2010).

Reply:
Ok, I guess it can stay like this until we implement more procedures.